chalice/chalicelib/matrix_handler.py

```diff
@@ -0,0 +1,172 @@
+
+        # Get all downloaded mtx paths from temp_dir
+        local_mtx_paths = get_mtx_paths(temp_dir, self._suffix)
+
+        logger.info("Done downloading %d matrix files.", len(local_mtx_paths))
+
+        return temp_dir, local_mtx_paths
+
+    @abstractmethod
+    def _concat_mtx(self, mtx_paths: List[str], mtx_dir: str, request_id: str) -> str:
+        """
+        Concatenate a list of matrices, and save into a new file on disk.
+
+        :param mtx_paths: A list of downloaded local matrix files paths.
+        :param mtx_dir: The directory that contains the matrices.
+        :param request_id: The request id for the matrices concatenation.
+        :return: New concatenated matrix file's path.
+        """
+
+    def _upload_mtx(self, path: str) -> str:
+        """
+        Upload a matrix file into an s3 bucket.
+        :param path: Path of the merged matrix.
+        :return: S3 bucket key for uploading file.
+        """
+        logger.info("%s", "Uploading \"{}\" to s3 bucket: \"{}\".".format(
+            os.path.basename(path),
+            MERGED_MTX_BUCKET_NAME
+        ))
+
+        key = os.path.basename(path)
+        with open(path, "rb") as merged_matrix:
+            s3_blob_store.upload_file_handle(
+                bucket=MERGED_MTX_BUCKET_NAME,
+                key=key,
+                src_file_handle=merged_matrix
+            )
+
+        logger.info("Done uploading.")
+
+        # Remove local merged mtx after uploading it to s3
+        shutil.rmtree(os.path.dirname(path))
+
+        return key
+
+    def run_merge_request(self, bundle_uuids: List[str], request_id: str, job_id: str) -> None:
+        """
+        Merge matrices within bundles, and upload the merged matrix to an s3 bucket.
+
+        :param bundle_uuids: Bundles' uuid for locating bundles in DSS.
+        :param request_id: Merge request id.
+        :param job_id: Job id of the request.
+        """
+        # Update the request status to RUNNING
+        try:
+            RequestHandler.update_request(
+                bundle_uuids=bundle_uuids,
+                request_id=request_id,
+                job_id=job_id,
+                status=RequestStatus.RUNNING
+            )
+
+            try:
+                start_time = time.time()
+                mtx_dir, mtx_paths = self._download_mtx(bundle_uuids)
+                merged_mtx_path = self._concat_mtx(mtx_paths, mtx_dir, request_id)
+                self._upload_mtx(merged_mtx_path)
+                end_time = time.time()
+
+                # Update the request status to DONE
+                RequestHandler.update_request(
+                    bundle_uuids=bundle_uuids,
+                    request_id=request_id,
+                    job_id=job_id,
+                    status=RequestStatus.DONE,
+                    time_spent_to_complete="{} seconds".format(end_time - start_time)
+                )
+            except (SwaggerAPIException, OSError, Exception):
```

Comment:
When timeout, the request message will be put back to the SQS queue, and waiting for a retry. After 3rd retry fails, it will be direct from the SQS Queue to a Dead Letter Queue. Right now, there is no further handling for the messages on the dead letter queue, but I intend to have some mechanism so that those requests will be set to "ABORT" status.

Reply:
Sounds like a plan, thanks.